qiskit/pulse/instructions/instruction.py

```diff
@@ -0,0 +1,286 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+``Instruction`` s are single operations within a :py:class:`~qiskit.pulse.Schedule`, and can be
+used the same way as :py:class:`~qiskit.pulse.Schedule` s.
+
+For example::
+
+    duration = 10
+    channel = DriveChannel(0)
+    sched = Schedule()
+    sched += Delay(duration, channel)  # Delay is a specific subclass of Instruction
+"""
+import warnings
+
+from abc import ABC
+
+from typing import Tuple, List, Iterable, Callable, Optional, Union
+
+from qiskit.pulse.channels import Channel
+from qiskit.pulse.interfaces import ScheduleComponent
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.timeslots import Interval, Timeslot, TimeslotCollection
+
+# pylint: disable=missing-return-doc
+
+# TODO: After migrating instruction implementations, add property+abstractmethod operands
+
+
+class Instruction(ScheduleComponent, ABC):
+    """The smallest schedulable unit: a single instruction. It has a fixed duration and specified
+    channels.
+    """
+
+    def __init__(self, duration: Union['Command', int],
+                 *channels: Channel,
+                 name: Optional[str] = None):
+        """Instruction initializer.
+
+        Args:
+            duration: Length of time taken by the instruction in terms of dt.
+            *channels: List of pulse channels that this instruction operates on.
+            name: Display name for this instruction.
+        """
+        self._name = name
+
+        self._command = None
+        if not isinstance(duration, int):
+            # TODO: Add deprecation warning once all instructions are migrated
+            self._command = duration
+            if name is None:
+                self._name = self.command.name
+            duration = self.command.duration
+
+        self._timeslots = TimeslotCollection(*(Timeslot(Interval(0, duration), channel)
+                                               for channel in channels if channel is not None))
```

Comment:
channels not being set is not necessarily a problem since this is an ABC, eg.,
```suggestion
                                               for channel in self.channels if channel is not None))
```

Reply:
Do we want to keep supporting the previous instructions that haven't been migrated yet, at least for a 3 month deprecation period? (e.g. DelayInstruction PulseInstruction FrameChangeInstruction etc)

Reply:
Yes, I think so as we cannot control the release interval and do not want to be caught in the middle of a migration. Once we have migrated do you think we can modify this to be as suggested?

Reply:
yes, I think so